runtime/v2/runc/task/service.go

```diff
@@ -517,48 +523,67 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-func (s *service) sendL(evt interface{}) {
-	s.eventSendMu.Lock()
-	s.events <- evt
-	s.eventSendMu.Unlock()
-}
-
 func (s *service) checkProcesses(e runcC.Exit) {
 	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	for _, container := range s.containers {
-		if !container.HasPid(e.Pid) {
-			continue
-		}
+	containers := s.containers
```

Comment:
Maps are reference types, similar to slices. [Copying a map by value only copies the reference to the underlying mutable data structure](https://go.dev/play/p/Wza0XD8b1MW) so this does not make it safe to access `containers` without holding `s.mu`. You would have to deep-copy the map to make this work.